rfcs/text/0352-graph-type-system.md

```diff
@@ -0,0 +1,2659 @@
+    "patternProperties": {
+      ".*": {
+        "type": "object",
+        "oneOf": [
+          {
+            "properties": {
+              "ordered": { "type": "boolean", "default": false },
+              "type": { "const": "array" }
+            },
+            "required": ["ordered", "type"]
+          },
+          {}
+        ],
+        "additionalProperties": false
+      }
+    }
+  }
+}
+```
+
+The `kind` can as explained in the Guide-Level explanation only take on values that specify the types of the proposed type system.
+
+The `links` keyword specifically allows constraints in the case of having a set of links. Here the link set can be ordered or unordered (default).
+
+As mentioned in the Guide-Level explanation, `description` for Link Types add semantic meaning to a link, which could be encapsulated in a vocabulary specification for the meta schemas.
+
+For the most part, we're using existing JSON Schema keywords, but it would be preferable to look into defining these additions as Vocabularies or otherwise integrate them into a JSON Schema validator used for the proposed system.
+
+## Using the Types in the Block Protocol
+
+Using the proposed type system for Block Protocol imposes changes on the Graph Service and how Block Schemas are defined.
+
+The Graph Service is currently specified as [this schema](https://github.com/blockprotocol/blockprotocol/blob/main/packages/%40blockprotocol/graph/src/graph-service.json).
+
+### Interfacing with properties on Entities
+
+A key change for allowing the proposed type system to work is moving away from arbitrary property keys and making use of canonical property URIs. As seen in the examples in earlier sections, properties that use simple keys will now have to point at _Property Type URIs_.
+
+**An example of an [`Entity`](https://blockprotocol.org/types/services/graph/entity) instance in the current system:**
+
+```json
+{
+  "entityId": 111,
+  "properties": {
+    "name": "Arthur Philip Dent",
+    "age": 30,
+    "planetOfOrigin": "Earth",
+    "occupation": "Intergalactic Traveler"
+  }
+}
+```
+
+**An example of an entity instance in the proposed system:**
+
+```json
+{
+  "https://blockprotocol.org/types/@blockprotocol/property-type/entity-id": 111,
+  "https://blockprotocol.org/types/@alice/property-type/name": "Arthur Philip Dent",
+  "https://blockprotocol.org/types/@alice/property-type/age": 30,
+  "https://blockprotocol.org/types/@alice/property-type/planet-of-origin": "Earth",
+  "https://blockprotocol.org/types/@alice/property-type/occupation": "Intergalactic Traveler"
+}
+```
+
+This change canonicalizes property keys, such that they uniquely identify a Property Type.
+
+**Within blocks** this impacts:
+
+- how entity fields are accessed
+- how Block Protocol messages are constructed
+- payload size of Block Protocol messages
+
+**For block authors** this impacts:
+
+- how Block Schemas are defined
+- reusability of properties (the need for some kind of type discovery)
```

Comment:
Can we unpack this? To the extent that BP users wanted to use other people's types in the current system, there would already be a need for some kind of type discovery, and this proposal doesn't increase the need to use other people's types (I suppose arguably out of convenience since it's now more work to define them?)

Reply:
```suggestion
- the reusability of properties and data types (and the increased need for type discovery to facilitate this)
```